DIPs/DIP4242.md

```diff
@@ -0,0 +1,260 @@
+either marks the delegate as `@safe`, and force the caller to use `@trusted` in some occasions,
+or avoid marking the function itself `@safe`, and not be callable from `@safe` code.
+In general, the former seems to be the prefered approach, as it makes the most sense.
+
+However, this choice is much less obvious for other attributes: forcing `nothrow`
+or `pure`ness is a less-accepted practice, let alone how restrictive forcing `@nogc` is.
+
+This problem can be seen in many widely used library, such as Vibe.d's delegate-accepting
+[`requestHTTP`](https://vibed.org/api/vibe.http.client/requestHTTP),
+or even druntime's [`Throwable.toString`](https://github.com/dlang/druntime/blob/d97ec4093b108dc2fa95f1fa04b1114e6e0611f8/src/object.d#L2020-L2026).
+It is also commonly seen when implementing `opApply`, as one usually has to choose between working type-inferencex
+(e.g. `foreach (varname; container)` as opposed to `foreach (type varname; container)`),
+which only works if the delegate type is known and not templated, or supporting multiple
+attributes, which is done by templating the delegate type.
+
+This proposal adds the ability to express the common dependency between a delegate's
+attributes and a function's attribute.
+
+## Prior Work
+
+[DIP1032](DIP1032.md) has proposed to always have the delegate parameters take the
+attributes of the function that receives it.
+However, this does not address the problem presented here: instead, it forces
+the user to require from the caller the attributes it supports.
+
+Doing so usually leads to attributes being faked, or the routine being unusable:
+for example, iterating over a collection might be a simple operation which is
+`pure` and `@nogc`, but requiring the delegate to also be is too limitating.
+
+## Description
+
+### Grammar
+
+The following changes to the grammar are proposed:
+```diff
+ StorageClass:
+     LinkageAttribute
+     AlignAttribute
+     deprecated
+     enum
+     static
+     extern
+     abstract
+     final
+     override
+     synchronized
+     auto
+     scope
+     const
+     immutable
+     inout
+     shared
+     __gshared
+     Property
+     nothrow
++    nothrow AttributeDeclDef
+     pure
++    pure AttributeDeclDef
+     ref
+ 
+ AtAttribute:
+     @ disable
+     @ nogc
++    @ nogc AttributeDeclDef
+     @ live
+     Property
+     @ safe
++    @ safe AttributeDeclDef
+     @ system
+     @ trusted
+     UserDefinedAttribute
+ 
++AttributeDeclDef:
++    ( * )
++    ( AttributeDeclDefs $(OPT) )
+
++AttributeDeclDefs:
++    AssignExpression
++    AssignExpression, AttributeDeclDefs ...
+```
+
+This syntax was choosen as it should feel familiar to the user,
+who can already encounter it when using UDAs.
+
+### Basic semantic rules
+
+If the argument list form is used, the argument must be either identifiers
+or integers. If identifiers, they must match the identifiers of one of
+the function's arguments, and this argument must be a delegate or function pointer.
```

Comment:
What about struct that define `opCall`?

Reply:
Currently you have to pass the `opCall`, but the DIP could be extended to support it.
However, if you do this, it's very likely the function is templated on the struct type.
One thing that comes to mind and could be quite interesting is to support it for `class` though:

```D
interface Callable { void opCall (); }
class Foo : Callable { override void opCall () nothrow {} }

void func (Callable c) @safe(*) {}
void main () nothrow
{
    Foo f = new Foo();
    func(f); // This could work
}
```

Reply:
Fair enough. The extension for classes sounds interesting though (even if it may rarely occur in practice).

---
 Anyway, this DIP should mention `opCall` and why it's currently excluded.